contracts/common/lib/DepositsTracker.sol

```diff
@@ -0,0 +1,203 @@
+    ///
+    /// @param _depositedEthStatePosition - slot in storage
+    /// @param currentSlot - slot of deposit
+    /// @param depositAmount - Eth deposit amount
+    function insertSlotDeposit(bytes32 _depositedEthStatePosition, uint256 currentSlot, uint256 depositAmount) public {
+        if (currentSlot > type(uint64).max) revert SlotTooLarge(currentSlot);
+        if (depositAmount > type(uint128).max) revert DepositAmountTooLarge(depositAmount);
+        // or maybe write this attempt to call tracker like we we call SR.deposit even if msg.value == 0
+        if (depositAmount == 0) revert ZeroValue("depositAmount");
+
+        DepositedEthState storage state = _getDataStorage(_depositedEthStatePosition);
+
+        uint256 depositsEntryAmount = state.slotsDeposits.length;
+
+        if (depositsEntryAmount == 0) {
+            state.slotsDeposits.push(SlotDeposit(uint64(currentSlot), uint192(depositAmount)).pack());
+
+            state.cursor = type(uint256).max;
+            return;
+        }
+
+        // last deposit
+        SlotDeposit memory lastDeposit = state.slotsDeposits[depositsEntryAmount - 1].unpack();
+
+        // if last tracked deposit's slot newer than currentSlot, than such attempt should be reverted
+        if (lastDeposit.slot > currentSlot) {
+            revert SlotOutOfOrder(lastDeposit.slot, currentSlot);
+        }
+
+        // if it is the same block, increase amount
+        if (lastDeposit.slot == currentSlot) {
+            lastDeposit.cumulativeEth += uint192(depositAmount);
+            state.slotsDeposits[depositsEntryAmount - 1] = lastDeposit.pack();
+
+            return;
+        }
+
+        state.slotsDeposits.push(
+            SlotDeposit(uint64(currentSlot), lastDeposit.cumulativeEth + uint192(depositAmount)).pack()
+        );
+    }
+
+    /// @notice Return the total ETH deposited before slot, inclusive slot
+    ///
+    /// @param _depositedEthStatePosition - slot in storage
+    /// @param _slot - Upper bound slot
+    /// @dev this method will use cursor for start reading data
+    function getDepositedEthUpToSlot(bytes32 _depositedEthStatePosition, uint256 _slot)
+        public
+        view
+        returns (uint256 total)
+    {
+        DepositedEthState storage state = _getDataStorage(_depositedEthStatePosition);
+        uint256 depositsEntryAmount = state.slotsDeposits.length;
+        if (depositsEntryAmount == 0) return 0;
+
+        // define cursor start
+        uint256 startIndex = 0;
+        uint256 leftBoundCumulativeSum = 0;
+
+        // if it was initialized earlier
+        if (state.cursor != type(uint256).max) {
+            if (state.cursor >= depositsEntryAmount) revert InvalidCursor(state.cursor, depositsEntryAmount);
+
+            SlotDeposit memory leftBoundDeposit = state.slotsDeposits[state.cursor].unpack();
+            if (leftBoundDeposit.slot > _slot) revert SlotOutOfRange(leftBoundDeposit.slot, _slot);
```

Comment:
what's the benefit of obtaining values in error parameters IRL? 
when debugging, the error signature clearly shows the location and cause.
imo, it's a waste of bytecode